File: static/js/ledger.js
function getLedgers(){
    return fetch('/api/credit')
        .then(response=>response.json())
}

function creditStatusChange(id){
    return fetch(`/api/creditstatus/${id}`)
        .then(response=>response.json())
}

function getAccount(){
    return fetch('/api/account')
        .then(response=>response.json())
}

getLedgers().then(myJson=>{
    myJson['data'].forEach((data)=>{
        const tr = document.createElement('tr');
        const id = document.createElement('td');
        const email = document.createElement('td');
        const bank = document.createElement('td');
        const amount = document.createElement('td');
        const blance = document.createElement('td');
        const act = document.createElement('td');
        const date = document.createElement('td');
        id.textContent=data['userId'];
        email.textContent = data['userEmail'];
        bank.textContent = data['userBank'];
        amount.textContent = data['amount'];
        blance.textContent = data['blance'];
        act.textContent = data['creditStatus']==false?'未付款':'已付款';
        let pay='';
        if(data['creditStatus']==false){
            pay=document.createElement('td');
            pay.className = 'pay-bt'
            pay.textContent='付款';
            pay.addEventListener('click',()=>{
                creditStatusChange(data['creditId']).then(myJson=>{
                    if(myJson['ok']){
                        location.reload();
                    }
                })
            })
        }
        date.textContent = data['date'];
        tr.append(id,email,bank,amount,blance,date,act,pay);
        document.getElementById('customers').append(tr);
    })
})

getAccount().then((data)=>{
    document.querySelector('#acc').textContent=data['amount']
})

File: api/ledger.py
import os
from flask import Blueprint, json,session,jsonify,redirect,request,make_response
import requests
from model.models import *
from cache import cache
import datetime


ledger = Blueprint('ledger',__name__)


@cache.memoize(60*30)
def calculate_balance(user):
    balance = 0
    for product in user.product:
        for order in product.order:
            balance+=order.total_price
    for withdraw in user.withdraw:
        balance-=withdraw.amount
    return balance


# @ledger.route('/api/debit',methods=['POST'])
# def deposit():
#     try:
#         rq = request.get_json()
#         debt = Debit(session['id'],rq['order_id'],rq['amount'],rq['receiver_id'])
#         db.session.add(debt)
#         db.session.commit()
#     except Exception as e:
#         print(e)
#         db.session.rollback()
#         return{
#             'error':True,
#         }



@ledger.route('/api/credit',methods=['POST'])
def withdraw():
    try:
        rq = request.get_json()
        credit = Credit(session['id'],rq['amount'])
        db.session.add(credit)
        db.session.commit()
        return{
            'ok':True
        }
    except Exception as e:
        print(e)
        db.session.rollback()
        return{
            'error':True,
        }

@ledger.route('/api/credit',methods=['GET'])
def get_ledger():
    if session['admin']!=True:
        er = {'error': True}
        return make_response(jsonify(er), 403)
    credit = db.session.query(Credit).all()
    datas = [{"creditStatus":data.activation,"creditId":data.id,"userId":data.user.id,"userEmail":data.user.email,"userBank":data.user.bank,"amount":data.amount,"blance":calculate_balance(db.session.query(User).filter_by(id=data.user_id).first()),"date":data.date} for data in credit]
    return {
        "data":datas
    }

@ledger.route('/api/account',methods=['GET'])
def get_account():
    if session['admin']!=True:
        er = {'error': True}
        return make_response(jsonify(er), 403)
    data = [order.total_price for order in db.session.query(Order).all()]
    credit_paid = [cdt.amount for cdt in db.session.query(Credit).filter_by(activation=True).all()]
    return{
        'amount':sum(data)-sum(credit_paid)
    }

@ledger.route('/api/creditstatus/<id>',methods=['GET'])
def pay_credit(id):
    if session['admin']!=True:
        er = {'error': True}
        return make_response(jsonify(er), 403)
    try:
        cdt = db.session.query(Credit).filter_by(id=id).first()
        cdt.activation=True
        db.session.commit()
        return{
            'ok':True
        }
    except Exception as e:
        print(e)
        db.session.rollback()
        return{
            'error':True
        }